naga/src/front/spv/mod.rs

```diff
@@ -1323,6 +1323,94 @@ impl<I: Iterator<Item = u32>> Frontend<I> {
+        log::trace!("\t\t\tlooking up pointer expr {:?}", pointer_id);
+        let (p_lexp_handle, p_lexp_ty_id) = {
+            let lexp = self.lookup_expression.lookup(pointer_id)?;
+            let handle = self.get_expr_handle(pointer_id, lexp, ctx, emitter, block, body_idx);
+            (handle, lexp.type_id)
+        };
+
+        log::trace!("\t\t\tlooking up pointer type {pointer_id:?}");
+        let p_ty = self.lookup_type.lookup(p_lexp_ty_id)?;
+        let p_ty_base_id = p_ty.base_id.ok_or(Error::InvalidAccessType(p_lexp_ty_id))?;
+
+        log::trace!("\t\t\tlooking up pointer base type {p_ty_base_id:?} of {p_ty:?}");
+        let p_base_ty = self.lookup_type.lookup(p_ty_base_id)?;
+
+        Ok((p_lexp_handle, p_base_ty.handle))
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn parse_atomic_expr_with_value(
+        &mut self,
+        inst: Instruction,
+        emitter: &mut crate::proc::Emitter,
+        ctx: &mut BlockContext,
+        block: &mut crate::Block,
+        block_id: spirv::Word,
+        body_idx: usize,
+        atomic_function: crate::AtomicFunction,
+    ) -> Result<(), Error> {
+        inst.expect(7)?;
+        let start = self.data_offset;
+        let span = self.span_from_with_op(start);
+        let result_type_id = self.next()?;
+        let result_id = self.next()?;
+        let pointer_id = self.next()?;
+        let _scope_id = self.next()?;
+        let _memory_semantics_id = self.next()?;
+        let value_id = self.next()?;
+
+        let (p_lexp_handle, p_base_ty_handle) =
+            self.get_exp_and_base_ty_handles(pointer_id, ctx, emitter, block, body_idx)?;
+
+        log::trace!("\t\t\tlooking up value expr {value_id:?}");
+        let v_lexp_handle = self.lookup_expression.lookup(value_id)?.handle;
+
+        block.extend(emitter.finish(ctx.expressions));
+        // Create an expression for our result
+        let r_lexp_handle = {
+            let expr = crate::Expression::AtomicResult {
+                ty: p_base_ty_handle,
+                comparison: false,
+            };
+            let handle = ctx.expressions.append(expr, span);
+            self.lookup_expression.insert(
+                result_id,
+                LookupExpression {
+                    handle,
+                    type_id: result_type_id,
+                    block_id,
+                },
+            );
+            handle
+        };
+        emitter.start(ctx.expressions);
+
+        // Create a statement for the op itself
+        let stmt = crate::Statement::Atomic {
+            pointer: p_lexp_handle,
+            fun: atomic_function,
+            value: v_lexp_handle,
+            result: Some(r_lexp_handle),
+        };
+        block.push(stmt, span);
+
+        // Store any associated global variables so we can upgrade their types later
+        self.upgrade_atomics
+            .extend(ctx.get_contained_global_variable(p_lexp_handle));
```

Comment:
Each time the SPV front end calls `get_contained_global_variable`, we use `extend` to add the global to `Frontend::upgrade_atomics`, This means that if `get_contained_global_variable` can't identify the global to which this atomic is referring, we don't upgrade anything. Is this the correct behavior? Why aren't these calls returning an error?

Reply:
Previously this function recursed and returned `IndexSet<Handle<crate::GlobalVariable>>`, but then I changed that to `Option<Handle<crate::GlobalVariable>>` as per your suggestion. But I think it makes sense to error, as we'll always be dealing with a global variable.

Reply:
Without pretending to understand the consequences, just in case, here's a commit that makes it an error if we can't identify a global to be upgraded: https://github.com/gfx-rs/wgpu/commit/52d43fd5c245001585d8a0253c8d845e76e6ba7b